hazelcast/src/main/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerRunner.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.management;
+
+import com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.spi.impl.ClientExecutionServiceImpl;
+import com.hazelcast.cluster.Address;
+import com.hazelcast.internal.util.executor.PoolExecutorThreadFactory;
+import com.hazelcast.logging.ILogger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Runs all the configured listeners on a specific executor, with proper
+ * error handling.
+ */
+public final class ClientConnectionProcessListenerRunner {
+
+    private final CopyOnWriteArrayList<ClientConnectionProcessListener> listeners;
+    private final HazelcastClientInstanceImpl client;
+    private final AtomicBoolean hasListeners;
+    private volatile ILogger logger;
+    private volatile ExecutorService executor;
```

Comment:
Asking to learn as I am still a Java dummy. Why do we use volatile for these? Telling the compiler not to cache these.

Reply:
Because the executor is initialized lazily, we couldn't make those fields final. So, they are written by, say, thread A, and possibly read by some other thread B, while firing the events. To make sure that the write done in thread A is visible to thread B or others, we need to make those fields volatile